xds/internal/xdsclient/clientimpl.go

```diff
@@ -100,3 +211,151 @@ func (c *clientImpl) close() {
 	}
 	c.logger.Infof("Shutdown")
 }
+
+func (c *clientImpl) getChannelForADS(serverConfig *bootstrap.ServerConfig, callingAuthority *authority) (*xdsChannel, func(), error) {
+	if c.done.HasFired() {
+		return nil, nil, ErrClientClosed
+	}
+
+	initLocked := func(s *channelState) {
+		if c.logger.V(2) {
+			c.logger.Infof("Adding authority %q to the set of interested authorities for channel [%p]", callingAuthority.name, s.channel)
+		}
+		s.interestedAuthorities[callingAuthority] = true
+	}
+	deInitLocked := func(s *channelState) {
+		if c.logger.V(2) {
+			c.logger.Infof("Removing authority %q from the set of interested authorities for channel [%p]", callingAuthority.name, s.channel)
+		}
+		delete(s.interestedAuthorities, callingAuthority)
+	}
+
+	return c.getOrCreateChannel(serverConfig, initLocked, deInitLocked)
+}
+
+func (c *clientImpl) getChannelForLRS(serverConfig *bootstrap.ServerConfig) (*xdsChannel, func(), error) {
+	if c.done.HasFired() {
+		return nil, nil, ErrClientClosed
+	}
+
+	initLocked := func(s *channelState) { s.lrsRefs++ }
+	deInitLocked := func(s *channelState) { s.lrsRefs-- }
+
+	return c.getOrCreateChannel(serverConfig, initLocked, deInitLocked)
+}
+
+func (c *clientImpl) getOrCreateChannel(serverConfig *bootstrap.ServerConfig, initLocked, deInitLocked func(*channelState)) (*xdsChannel, func(), error) {
+	c.channelsMu.Lock()
+	defer c.channelsMu.Unlock()
+
+	if c.logger.V(2) {
+		c.logger.Infof("Received request for a reference to an xdsChannel for server config %q", serverConfig)
+	}
+
+	// Use an active channel, if one exists for this server config.
+	if state, ok := c.xdsActiveChannels[serverConfig.String()]; ok {
+		if c.logger.V(2) {
+			c.logger.Infof("Reusing an active xdsChannel for server config %q", serverConfig)
+		}
+		initLocked(state)
+		return state.channel, c.releaseChannel(serverConfig, state, deInitLocked), nil
+	}
+
+	// If an idle channel exists for this server config, remove it from the
+	// idle cache and add it to the map of active channels, add return it.
+	if state, ok := c.xdsIdleChannels.Remove(serverConfig.String()); ok {
+		if c.logger.V(2) {
+			c.logger.Infof("Reviving an xdsChannel from the idle cache for server config %q", serverConfig)
+		}
+		state, _ := state.(*channelState)
+		c.xdsActiveChannels[serverConfig.String()] = state
```

Comment:
`c.xdsActiveChannels[serverConfig.String()] = state`. Shouldn't this need to be synchronized too like inside lock/unlock critical section? Actually may be the whole if block  starting 266 needs to be in critical section?

Reply:
I grab the lock at the top of this method. Are you saying that is not enough?